File: gamelogic/src/main/java/com/alta_v2/game/gamelogic/stage/AbstractStage.java
package com.alta_v2.game.gamelogic.stage;

import com.alta_v2.game.gamelogic.stage.event.ChangeMapEventHandler;
import com.alta_v2.game.gamelogic.stage.event.ChangeStageEvent;
import com.alta_v2.game.gamelogic.stage.event.ChangeScreenHandler;
import lombok.extern.log4j.Log4j2;

@Log4j2
public abstract class AbstractStage implements Stage {

    private ChangeScreenHandler changeScreenHandler;
    private ChangeMapEventHandler changeMapHandler;

    /**
     * {@inheritDoc}
     */
    @Override
    public void subscribeToChangeScreen(ChangeScreenHandler handler) {
        this.changeScreenHandler = handler;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void subscribeToChangeMap(ChangeMapEventHandler handler) {
        this.changeMapHandler = handler;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void destroy() {
        this.changeMapHandler = null;
        this.changeScreenHandler = null;
    }

    protected void changeStage(ChangeStageEvent changeStageEvent) {
        if (this.changeScreenHandler == null) {
            log.info("Change screen handler is null.");
            return;
        }

        this.changeScreenHandler.handle(changeStageEvent);
    }
}

File: gamelogic/src/main/kotlin/com/alta_v2/game/gamelogic/stage/AbstractStage.kt
package com.alta_v2.game.gamelogic.stage

import com.alta_v2.game.gamelogic.stage.event.ChangeMapEventHandler
import com.alta_v2.game.gamelogic.stage.event.ChangeScreenHandler
import com.alta_v2.game.gamelogic.stage.event.ChangeStageEvent
import mu.KotlinLogging

private val log = KotlinLogging.logger {  }

abstract class AbstractStage : Stage {

    private var changeScreenHandler: ChangeScreenHandler? = null
    private var changeMapHandler: ChangeMapEventHandler? = null

    /**
     * {@inheritDoc}
     */
    override fun subscribeToChangeScreen(handler: ChangeScreenHandler) {
        this.changeScreenHandler = handler
    }

    /**
     * {@inheritDoc}
     */
    override fun subscribeToChangeMap(handler: ChangeMapEventHandler) {
        changeMapHandler = handler
    }

    /**
     * {@inheritDoc}
     */
    override fun onStageLoaded() {}

    /**
     * {@inheritDoc}
     */
    override fun destroy() {
        changeMapHandler = null
        changeScreenHandler = null
    }

    protected open fun changeStage(changeStageEvent: ChangeStageEvent) {
        if (changeScreenHandler == null) {
            log.info("Change screen handler is null.")
            return
        }
        changeScreenHandler!!.handle(changeStageEvent)
    }

}